src/titan_checkpoint_impl.cc

```diff
@@ -0,0 +1,300 @@
+            return CreateFile(db_->GetEnv(), full_private_path + fname,
+                              contents, db_options.use_fsync);
+          } /* create_file_cb */,
+          &sequence_number, log_size_for_flush, full_private_path);
+    }
+
+    if (disabled_file_deletions) {
+      // We copied all the files, enable file deletions
+      Status ss = db_->EnableFileDeletions(false);
+      assert(ss.ok());
+    }
+  }
+
+  if (s.ok()) {
+    // Move tmp private backup to real checkpoint directory
+    s = db_->GetEnv()->RenameFile(full_private_path, checkpoint_dir);
+  }
+  if (s.ok()) {
+    std::unique_ptr<Directory> checkpoint_directory;
+    db_->GetEnv()->NewDirectory(checkpoint_dir, &checkpoint_directory);
+    if (checkpoint_directory != nullptr) {
+      s = checkpoint_directory->Fsync();
+    }
+  }
+
+  if (s.ok()) {
+    // Here we know that we succeeded and installed the new checkpoint
+    ROCKS_LOG_INFO(db_options.info_log, "Checkpoint DONE. All is good");
+    ROCKS_LOG_INFO(db_options.info_log, "Checkpoint sequence number: %" PRIu64,
+                   sequence_number);
+  } else {
+    // Clean all the files we might have created
+    ROCKS_LOG_INFO(db_options.info_log, "Checkpoint failed -- %s",
+                   s.ToString().c_str());
+    CleanStagingDirectory(full_private_path, db_options.info_log.get());
+  }
+  return s;
+}
+
+Status TitanCheckpointImpl::CreateCustomCheckpoint(
+    const DBOptions& db_options,
+    std::function<Status(const std::string& src_dirname,
+                         const std::string& src_fname, FileType type)>
+        link_file_cb,
+    std::function<Status(const std::string& src_dirname,
+                         const std::string& src_fname,
+                         uint64_t size_limit_bytes, FileType type)>
+        copy_file_cb,
+    std::function<Status(const std::string& fname, const std::string& contents,
+                         FileType type)>
+        create_file_cb,
+    uint64_t* sequence_number, uint64_t log_size_for_flush,
+    const std::string full_private_path) {
+  Status s;
+  std::vector<std::string> titandb_files;
+  std::vector<VersionEdit> version_edits;
+  bool same_fs = true;
+
+  // Create base db checkpoint
+  s = db_->DisableFileDeletions();
+  const bool disabled_file_deletions = s.ok();
+
+  auto base_db_checkpoint = new rocksdb::CheckpointImpl(db_);
+  s = base_db_checkpoint->CreateCustomCheckpoint(
+      db_options, link_file_cb, copy_file_cb, create_file_cb, sequence_number,
+      log_size_for_flush);
+  delete base_db_checkpoint;
+  base_db_checkpoint = nullptr;
+
+  if (!s.ok()) {
+    if (disabled_file_deletions) {
+      Status ss = db_->EnableFileDeletions(false);
+      assert(ss.ok());
+    }
+    return s;
+  }
+
+  // This will return files prefixed with "/titandb"
+  s = db_->GetAllTitanFiles(titandb_files, &version_edits);
+
+  if (disabled_file_deletions) {
+    Status ss = db_->EnableFileDeletions(false);
```

Comment:
This should be called after linking or copying the files. And I check the RocksDB's code, seems it just calls `EnableFileDeletions` and `DisableFileDeletions` in `CreateCheckpoint`. So let's just remove these(sorry about that)

Reply:
Ok, I'll change it right away.